diary/ads-support.org

```diff
@@ -0,0 +1,62 @@
+#+TITLE: Ads Support
+#+DATE: 2021-01-18
+#+AUTHOR: Zach Mandeville
+
+** Introduction
+This diary is a work-in-progress, documenting how we add ADS support to the xDS
+test harness. It starts by checking assumptions I have about ADS, sets out an
+initial design for the work, and then acts as a work diary of my pgoress.
+
+The bottom of the document has open questions for which I'd love feedback. When
+a question is answered, it'll be removed from that section and its answer woven
+into this diary.
+
+** Assumptions/Current Understanding
+- The xDS transport protocol has [[https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#variants-of-the-xds-transport-protocol][four variants]]. An xDS server can handle any or
+  all of these variants.
+- The test framework, too, should cover all variants.
+  + It [[https://github.com/ii/xds-test-harness/tree/de750b5ba26ba3f0de5a6ecca1fcac20b787ee30][currently]] only supports SoTW(State of the World)/Basic
+  + SoTW/ADS(Aggregated Discovery) should be the added next, followed by Delta/Basic and Delta/ADS.
+- An xDS server is configured for a single variant on startup
+  + e.g: if it's set to SoTW/ADS, then all CDS,LDS,RDS,EDS features will be handled
+    via ADS
+  + It cannot switch itself from basic to ADS and remain running.
+- The intent of our current tests are the same, no matter if a server is
+  configured as basic or aggregated.
+  + However, there will be additional 'ADS-only' tests
+- A team may want to ensure their server can handle basic /and/ ADS.
```

Comment:
A given server may support any arbitrary subset of the protocol variants.  Some possible examples of what a server might support:
- non-aggregated SotW only
- SotW only, both aggregated and non-aggregated (but neither of the incremental variants)
- aggregated only, both SotW and incremental (but neither of the non-aggregated variants)
- all 4 variants
- only aggregated SotW and non-aggregated incremental (but not non-aggregated SotW or aggregated incremental)

That last example is probably fairly unlikely, but there's no reason that the test framework should make any assumptions about it.